src/childinfo.c

```diff
@@ -54,34 +60,82 @@ void closeChildInfoPipe(void) {
     if (server.child_info_pipe[1] == -1) return;
-    server.child_info_data.magic = CHILD_INFO_MAGIC;
-    server.child_info_data.process_type = ptype;
-    ssize_t wlen = sizeof(server.child_info_data);
-    if (write(server.child_info_pipe[1],&server.child_info_data,wlen) != wlen) {
+
+    child_info_data buffer = {.process_type = process_type, .cow_size = cow_size};
+    ssize_t wlen = sizeof(buffer);
+
+    if (write(server.child_info_pipe[1],&buffer,wlen) != wlen) {
         /* Nothing to do on error, this will be detected by the other side. */
     }
 }
 
+/* Update COW data. */
+void updateChildInfo(int process_type, size_t cow_size) {
+    if (process_type == CHILD_TYPE_RDB) {
+        server.stat_rdb_cow_bytes = cow_size;
+    } else if (process_type == CHILD_TYPE_AOF) {
+        server.stat_aof_cow_bytes = cow_size;
+    } else if (process_type == CHILD_TYPE_MODULE) {
+        server.stat_module_cow_bytes = cow_size;
+    }
+}
+
+/* Read COW data */
+int readChildInfo(int *process_type, size_t *cow_size) {
+    /* static buffer to read data from the pipe */
+    static child_info_data buffer;
+    ssize_t wlen = sizeof(buffer);
+
+    /* do not overlap */
+    if (server.child_info_nread == wlen) server.child_info_nread = 0;
+
+    int nread = read(server.child_info_pipe[0],&buffer + server.child_info_nread, wlen - server.child_info_nread);
+    if (nread > 0) {
+        server.child_info_nread += nread;
+    }
+
+    /* we have complete child info */
+    if (server.child_info_nread == wlen) {
+        *process_type = buffer.process_type;
+        *cow_size = buffer.cow_size;
+
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
 /* Receive COW data from parent. */
 void receiveChildInfo(void) {
     if (server.child_info_pipe[0] == -1) return;
-    ssize_t wlen = sizeof(server.child_info_data);
-    if (read(server.child_info_pipe[0],&server.child_info_data,wlen) == wlen &&
-        server.child_info_data.magic == CHILD_INFO_MAGIC)
-    {
-        if (server.child_info_data.process_type == CHILD_TYPE_RDB) {
-            server.stat_rdb_cow_bytes = server.child_info_data.cow_size;
-        } else if (server.child_info_data.process_type == CHILD_TYPE_AOF) {
-            server.stat_aof_cow_bytes = server.child_info_data.cow_size;
-        } else if (server.child_info_data.process_type == CHILD_TYPE_MODULE) {
-            server.stat_module_cow_bytes = server.child_info_data.cow_size;
-        }
+
+    int process_type;
+    size_t cow_size;
+
+    if (readChildInfo(&process_type, &cow_size)) {
+        updateChildInfo(process_type, cow_size);
+    }
+}
+
+/* Receive last COW data from parent. */
+void receiveLastChildInfo(void) {
+    if (server.child_info_pipe[0] == -1) return;
+
+    /* drain the pipe, and check last read data (if exist) */
+    int child_info_available = 0;
+    int process_type;
+    size_t cow_size;
+
+    while (readChildInfo(&process_type, &cow_size) > 0) child_info_available = 1;
```

Comment:
nitpick, i would have moved the body of the loop to a line of itself, and even added a spare line for braces.
i think loop need to stand out (be more visible).
unless there's some error handling pattern or something that's repeated many times.